### PR TITLE
Add support for getting custom payloads on executes

### PR DESCRIPTION
**Pull request by maxdymond**
Signed-off-by: Max Dymond <max.dymond@metaswitch.com>

### CODE REVIEW

File: floki.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+image:
```
Comment (kw217) [inline floki.yaml:1]: Great idea! Please could you add some documentation on how to use floki (and a link to it) to the README?

File: src/cassandra/batch.rs [hunk 2]
```diff
@@ -43,50 +41,6 @@ impl Protected<*mut _Batch> for Batch {
     }
 }
 
-/// Custom payloads not fully supported yet
```
Comment (kw217) [inline src/cassandra/batch.rs:46]: Please can you update CHANGELOG to reflect your changes in this PR?

File: src/cassandra/custom_payload.rs [hunk 3]
```diff
@@ -1,7 +1,56 @@
...
+use crate::cassandra_sys::cass_custom_payload_new;
+use crate::cassandra_sys::cass_custom_payload_set;
+use crate::cassandra_sys::cass_custom_payload_set_n;
+use crate::cassandra_sys::CassCustomPayload as _CassCustomPayload;
+use std::collections::HashMap;
+use std::os::raw::c_char;
+
+pub type CustomPayloadResponse = HashMap<String, Vec<u8>>;
+
+/// Custom payloads not fully supported yet
```
Comment (kw217) [inline src/cassandra/custom_payload.rs:14]: Is this comment still true?

File: src/cassandra/custom_payload.rs [hunk 4]
```diff
@@ -1,7 +1,56 @@
...
+#[derive(Debug)]
+pub struct CustomPayload(*mut _CassCustomPayload);
+
```
Comment (kw217) [inline src/cassandra/custom_payload.rs:17]: I think you can probably make this `Send`, though obviously not `Sync` - see justification elsewhere e.g. https://github.com/Metaswitch/cassandra-rs/blob/master/src/cassandra/collection.rs#L432 .

File: src/cassandra/future.rs [hunk 5]
```diff
@@ -123,6 +127,54 @@ impl Completable for PreparedStatement {
...
+            .to_result((name, name_length, value, value_size))
+            .and_then(|(name, name_length, value, value_size)| {
+                let name_slice = slice::from_raw_parts(name as *const u8, name_length);
+                str::from_utf8(name_slice)
+                    .map_err(|err| err.into())
+                    .map(|name| {
+                        (
+                            name.to_string(),
+                            slice::from_raw_parts(value, value_size).to_vec(),
+                        )
```
Comment (kw217) [inline src/cassandra/future.rs:160]: Nice. I confirm I've reviewed the C++ docs and this code and it looks to me like you're doing all the right things.

File: src/cassandra/future.rs [hunk 6]
```diff
@@ -123,6 +127,54 @@ impl Completable for PreparedStatement {
...
+                    })
+            })
+        })
+        .collect::<Result<CustomPayloadResponse>>()
+}
+
+/// Futures that complete with a normal result and a custom payload response.
```
Comment (kw217) [inline src/cassandra/future.rs:167]: Nice!

File: src/cassandra/session.rs [hunk 7]
```diff
@@ -135,6 +137,19 @@ impl Session {
         unsafe { <CassFuture<CassResult>>::build(cass_session_execute(self.0, statement.inner())) }
     }
 
+    /// Execute a statement and get any custom payloads.
```
Comment (kw217) [inline src/cassandra/session.rs:140]: Please can you implement `execute_batch_with_payloads` as well?

Comment (kw217) [inline src/cassandra/session.rs:140]: Can you elaborate slightly in the comment, and say "get any custom payloads _from the response_"?

File: src/cassandra/value.rs [hunk 8]
```diff
@@ -306,10 +306,9 @@ impl Value {
         unsafe {
             let mut output = mem::zeroed();
             let mut output_size = mem::zeroed();
-            let result = cass_value_get_bytes(self.0, &mut output, &mut output_size);
-            // raw2utf8(output, output_size).unwrap()
-            let slice = slice::from_raw_parts(output, output_size as usize);
-            result.to_result(slice)
+            cass_value_get_bytes(self.0, &mut output, &mut output_size)
+                .to_result((output, output_size))
+                .map(|(output, output_size)| slice::from_raw_parts(output, output_size as usize))
```
Comment (kw217) [inline src/cassandra/value.rs:311]: :-) thanks.

**Review by kw217 — CHANGES_REQUESTED** [review]
Thanks - looking good! I've requested some docs, and a couple of tiny tweaks. Thank you very much for this contribution - it's going to be very useful.

File: floki.yaml (hunk 1 above)
Comment (maxdymond) [reply floki.yaml:1]: Yes, will do.

File: src/cassandra/batch.rs (hunk 2 above)
Comment (maxdymond) [reply src/cassandra/batch.rs:46]: Yes.

File: src/cassandra/custom_payload.rs (hunk 3 above)
Comment (maxdymond) [reply src/cassandra/custom_payload.rs:14]: Great question - I don't think this adds support for adding custom payloads to queries yet, so yes I think it is still true.

File: src/cassandra/custom_payload.rs (hunk 4 above)
Comment (maxdymond) [reply src/cassandra/custom_payload.rs:17]: I just moved the code from batch.rs; do you want me to do this?

File: src/cassandra/future.rs (hunk 5 above)
Comment (maxdymond) [reply src/cassandra/future.rs:160]: 👍

File: src/cassandra/future.rs (hunk 6 above)
Comment (maxdymond) [reply src/cassandra/future.rs:167]: 👍

File: src/cassandra/session.rs (hunk 7 above)
Comment (maxdymond) [reply src/cassandra/session.rs:140]: Yes, sure.

Comment (maxdymond) [reply src/cassandra/session.rs:140]: Yes

File: src/cassandra/value.rs (hunk 8 above)
Comment (maxdymond) [reply src/cassandra/value.rs:311]: 👍

**Comment by maxdymond** [comment]
@kw217 please check the latest commit; I haven't explicitly changed the `Send` code as per my comment above, but should have fixed everything else.

File: src/cassandra/custom_payload.rs (hunk 4 above)
Comment (kw217) [reply src/cassandra/custom_payload.rs:17]: Yes please. If we're going to start using these, we're going to want them to play nicely with futures-0.3.

Comment (maxdymond) [reply src/cassandra/custom_payload.rs:17]: Fixed.

**Review by kw217 — APPROVED** [review]

**Comment by kw217** [comment]
LGTM - thanks @maxdymond !

**Pull request merged**